### DISCUSSION TITLE
Support code exclusion from coverage instrumentation

### DISCUSSION THREAD

**Discussion by jozic** [scala/scala3#17685] [Feature Requests]
Original scalac scoverage plugin supports excluding code from instrumentation and reports
https://github.com/scoverage/scalac-scoverage-plugin#excluding-code-from-coverage-stats
Scala 3 can support it as well, to make transition from Scala 2 easier.

Also, any hints for volunteers on where to start are highly appreciated :)
Thank you!

**Comment by TheElectronWill** [comment]
Okay, here's some hint :)
Coverage instrumentation is done by the phase `instrumentCoverage` in `compiler/src/dotty/tools/dotc/transform/InstrumentCoverage.scala`. 

The most performant form of exclusion is to exclude files. The current "compilation unit" provides the source that is being compiled:
```scala
ctx.compilationUnit.source
```

This could be used to disable the instrumentation phase according to a collection of regex-like strings. We have `-coverage-out` to configure the output of scoverage files, we could have `-coverage-exclude-files` or something like that. Such options are defined in `ScalaSettings`, you can see how `-coverage-out` has been defined.

Excluding packages (or worse, classes) looks a bit more complicated.

**Comment by tOverney** [comment]
It would be great to have the coverage exclusion supported in Scala3;
We mainly rely on it to remove any instrumentation from generated classes in src_managed.
We do this with `coverageExcludedPackages`

**Comment by george-wilson-rea** [comment]
I would love to see coverage checking exclusion for Scala 3. This is the last thing that is preventing me from upgrading many projects to Scala 3.

**Comment by jtjeferreira** [comment]
I would like to add that any ignoring mechanism should also support `// $COVERAGE-OFF$` comments...

**Reply by TonioGela** [reply]
@TheElectronWill do you know how also this feature can be supported by the compiler?

**Reply by TheElectronWill** [reply]
This one could be more difficult to implement because it would require to modify the Parser and to pass down the exclusion information to the InstrumentCoverage phase. IMO an annotation would be easier (because annotations are already taken care of by the Parser and won't require a special case for the comments).

**Reply by TonioGela** [reply]
I see, thanks for the pointer

**Reply by jozic** [reply]
Annotations can be easier to implement but harder to use. Comments allow to exclude anything, while annotations can be applied to elements only afaik. So comments would be a preferred solution for users (at least for me :)), especially given that it works like that for Scala 2 for years. Cross compiling also would be easier.

Having said that I do understand we need to find a balance.

**Reply by james-work-account** [reply]
Do you know if there's been any movement on supporting excluding lines with either comments or annotations? I am working through upgrading some Play! Framework services, and their default implementation of JSON serialisation/deserialisation - `implicit val format: OFormat[MyClass] = Json.format[MyClass]` - is no longer being ignored like it was in Scala 2.13. It's meaning that repositories which previously had 100% coverage are now significantly lower, and not being able to exclude specific lines of code is making the job, of getting coverage back to where it was, frustrating.

**Reply by jducoeur** [reply]
Just another data point, but this is a *big* roadblock for my team's attempt to migrate towards Scala 3.  We've had a firm rule of "100%" coverage for many years, and have adhered to that well -- but that "100%" absolutely requires being able to do fine-grained exclusions of stuff that we know can't be tested in any practical way.  (Or which we are intentionally excluding from coverage for a short time, until a ticket or some such to rectify it.)

Without being able to nail coverage to 100% in the build, you're stuck with ad-hockery, which is bad for any sort of serious software.  (I have [a whole post on just this subject](https://jducoeur.medium.com/a-philosophy-of-testing-5-the-case-for-100-a0ddea055002), if anyone cares.)

**Reply by tgodzik** [reply]
We should finish https://github.com/scala/scala3/pull/24486 
There exclude flags are already available otherwise. Is there anything more we should add?

**Reply by jozic** [reply]
I believe other exclusion settings are already supported

**Comment by maslovalex** [comment]
Does anybody knows if https://github.com/scala/scala3/pull/19727 planned to be back-ported to LTS branch ?
